### DISCUSSION TITLE
Add "focus-follows-mouse"-like option but for outputs only

### DISCUSSION THREAD

**Discussion by xezo360hye** [YaLTeR/niri#378] [Ideas]
I like how "focus follows mouse" works in general but it's not something I could normally use here with two monitors placed side-by-side since moving mouse to the other screen can move the windows view frame. Therefore I have disabled it, but it's somewhat annoying that I need to click to focus even when going with mouse to the other monitor. I wish there was an option like "focus-follows-mouse" but exclusively for outputs.

Since focused window on each monitor is already stored separately (I assume so since there exists "focus-ring") it shouldn't require a lot of changes in existing code I guess. A general algorithm is the following:

1. For each mouse movement event
2. Check whether the pointer has moved off screen
3. Set the screen under cursor as focused
4. Set last focused window on it as currently focused

Didn't look into the code yet, maybe the 4th step isn't needed at all. I'd try to implement it myself but I don't have enough time for it as of now, and at least not until exams are over.

**Comment by YaLTeR** [comment]
Hmm, maybe, yeah. You're right, every monitor (every workspace actually) stores a focused window, there's no need to do anything extra.

**Comment by tv42** [comment]
Somewhat related, I would love a focus-follows-mouse that changes focus (even within one output), but never changes the layout. On niri 0.1.6, if you have a screen that shows

```
<--screen width-->
| 1/2 col | full width col |
```

and you move the pointer from the left column to the right column, the screen scrolls to only show the right column.

I don't want scrolling to happen just because I'm moving the pointer, but I can't live without focus-follows-mouse.

**Reply by YaLTeR** [reply]
The focused window is always scrolled into view, regardless of how you focused it. There are some special cases like "don't move the view during an interactive resize". But with focus-follows-mouse there's no case to special case, it just changes the focus. So I'm not sure how this could work.

**Reply by tv42** [reply]
Would it take anything more than making that a configuration boolean? Setting that to false, the user can scroll columns when they want to. When true, after every focus change it brings the column/window fully into view.

**Reply by YaLTeR** [reply]
Do you mean that with false, the columns will never scroll on their own? I think that would get very annoying very quickly

**Reply by tv42** [reply]
Yes. I *need* focus-follows-mouse but IMHO it's jolting to see the windows jump around from just mouse movement. 

Consider a floating window manager: if I move a window partially off-screen, it doesn't jump into view when I hover over it.

**Reply by YaLTeR** [reply]
Well, in a floating window manager you don't frequently have windows mostly off-screen as part of the normal use.

**Reply by tv42** [reply]
Here's another scenario in which the windows moving around bothers me:

A monitor is on the right of the laptop. The monitor has a leftmost column that is not fully visible. I've been interacting with the two 1/3rd width columns, and want to leave them visible to keep reading their content. I move the mouse pointer from the middle column of the monitor to the laptop display, to write notes there.

The current Niri will scroll the columns on the monitor, making only the full width window visible, hiding the windows I wanted to refer to.

If I could disable this automatic scrolling, the columns on the monitor would remain visible.

```
<--laptop-->
| whatever |
            <-----------monitor----------->
 |..   full width | one third | one third |
```

**Reply by xezo360hye** [reply]
> Here's another scenario in which the windows moving around bothers me:
> 
> 
> 
> A monitor is on the right of the laptop. The monitor has a leftmost column that is not fully visible. I've been interacting with the two 1/3rd width columns, and want to leave them visible to keep reading their content. I move the mouse pointer from the middle column of the monitor to the laptop display, to write notes there.

In fact you just described my exact setup and the reason I created this issue.

Btw I'd appreciate if someone could point me to the relevant files/directories of the project now that I have a bit more time.

**Reply by YaLTeR** [reply]
The config preference: https://github.com/YaLTeR/niri/blob/22a948cc759498923c1e9806580962cdd2d7b3e2/niri-config/src/lib.rs#L79

Since we need to retain backwards compatibility, it should still work as on/off by itself, so maybe a good way would be to add a property to it, i.e. `focus-follows-mouse outputs-only=true`. This may need wrangling with knuffel to do custom parsing. There are various custom parsing examples with other parts of the config.

The logic itself: should be trivially changeable here https://github.com/YaLTeR/niri/blob/22a948cc759498923c1e9806580962cdd2d7b3e2/src/niri.rs#L3828

**Comment by YaLTeR** [comment]
Let's discuss in https://github.com/YaLTeR/niri/discussions/491
